### PR TITLE
check jenkins result state

### PR DESCRIPTION
**Pull request by tnyz**
Allow user to specify target states, add connection schema support in jenkins url
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by eladkal** [comment]
Is this related to https://github.com/apache/airflow/issues/28465 ?

**Comment by tnyz** [comment]
No, I think what was described in that ticket is about scheme instead of schema `scheme://user:password@host%2Flocation:1234/schema?__extra__=%7B%22my_val%22%3A+%5B%22list%22%2C+%22of%22%2C+%22values%22%5D%2`

**Comment by tnyz** [comment]
@eladkal any chance you can review this? thanks

File: airflow/providers/jenkins/hooks/jenkins.py [hunk 1]
```diff
@@ -40,21 +39,34 @@ def __init__(self, conn_id: str = default_conn_name) -> None:
...
         """Get jenkins server"""
         return self.jenkins_server
 
+    def get_latest_build_number(self, job_name) -> int:
+        try:
+            self.log.info("Build number not specified, getting latest build info from Jenkins")
+            job_info = self.jenkins_server.get_job_info(job_name)
+            return job_info["lastBuild"]["number"]
+        except jenkins.JenkinsException as err:
+            raise AirflowException(f"Jenkins call failed with error : {err}")
```
Comment (uranusjr) [inline airflow/providers/jenkins/hooks/jenkins.py]: Do not catch this, let the Jenkins error bubble. (Same for other functions.)

File: airflow/providers/jenkins/sensors/jenkins.py [hunk 2]
```diff
@@ -42,21 +43,28 @@ def __init__(
...
             return False
-        else:
-            self.log.info("Build is finished.")
+
+        build_result = hook.get_build_result(self.job_name, build_number)
+        self.log.info(f"Build is finished, result is {build_result}.")
+        if build_result in self.target_states:
             return True
+        else:
+            raise AirflowException("The build result does not meet the target states.")
```
Comment (uranusjr) [inline airflow/providers/jenkins/sensors/jenkins.py]: Can this be more specific? Can the user do something to fix this, or is this an impossible situation the user should tell the Jenkins admin?

**Review by uranusjr — CHANGES_REQUESTED** [review]
Also you need to fix CI errors.

File: tests/providers/jenkins/sensors/test_jenkins.py [hunk 3]
```diff
@@ -21,37 +21,34 @@
...
 
 class TestJenkinsBuildSensor:
     @pytest.mark.parametrize(
-        "build_number, build_state",
+        "build_number, build_state, result",
         [
-            (
-                1,
-                False,
-            ),
```
Comment (uranusjr) [inline tests/providers/jenkins/sensors/test_jenkins.py:35]: Why is this case removed?

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Need to fix static checks.

**Comment by tnyz** [comment]
> Need to fix static checks.

Good to merge

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Pull request merged**